### PR TITLE
binary parser for vw

### PR DESCRIPTION
**Pull request by olgavrou**
paired with https://github.com/VowpalWabbit/vowpal_wabbit/pull/2829

will update vw submodule before merging

build completely independently from rlclientlib:

```
cd external_parser
mkdir build
cd build
cmake ..
make
```

vw executable should then be found at: `external_parser/build/vw_binary_parser/vowpalwabbit/vw`

run using `--binary_parser` flag

### CODE REVIEW

File: external_parser/parse_example_external.cc [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+  auto len = all->example_parser->input->buf_read(line, sizeof(payload_type));
+
+  if (len < sizeof(payload_type)) {
+    return false;
+  }
+
+  payload_type = *reinterpret_cast<const unsigned int *>(line);
+  if (payload_type == 0) {
+    // try again?
```
Comment (olgavrou) [inline external_parser/parse_example_external.cc:116]: this happens because the python parser adds padding

File: CMakeLists.txt [hunk 2]
```diff
@@ -31,18 +32,21 @@ ProcessorCount(NumProcessors)
 message("Number of processors: ${NumProcessors}")
 file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/nprocs.txt ${NumProcessors})
```
Comment (olgavrou) [inline CMakeLists.txt:34]: Not sure if this should be built this way or if the external parser should be standalone and built from inside the `external_parser` directory

File: external_parser/example_joiner.cc [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+    if (metadata->payload_type() == v2::PayloadType_Outcome) {
+      process_outcome(*event, *metadata);
+    } else {
+      process_interaction(*event, *metadata, examples);
+    }
+  }
+  // call logic that creates the reward
+  reward_calculation(_unjoined_examples[id], examples);
+  // return an empty example to signal end-of-multiline
```
Comment (olgavrou) [inline external_parser/example_joiner.cc:152]: oh need to add check here that we are actually multi-line before adding an empty example

File: CMakeLists.txt [hunk 4]
```diff
@@ -31,18 +32,21 @@ ProcessorCount(NumProcessors)
...
-add_subdirectory(test_tools/joiner)
-add_subdirectory(test_tools/sender_test)
-add_subdirectory(test_tools/example_gen)
-add_subdirectory(test_tools/log_parser)
-
-# enable_testing should be run after ext_libs so that the vw unit tests arent turned on.
-enable_testing()
-add_subdirectory(unit_test)
-add_subdirectory(unit_test/extensions)
+if(BUILD_VW_EXTERNAL_PARSER)
```
[the remark this answers is not in the document]
Comment (olgavrou) [reply CMakeLists.txt]: I think you are looking at an outdated version of the PR

**Review by kumpera — APPROVED** [review]
LGTM

**Pull request merged**